### ISSUE TITLE
autograd.functional.vjp with function that mutates inputs in-place can cause confusion

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#52390]
## 🐛 Bug

Define a function that mutates its first argument in-place. Call `F.vjp` with `create_graph=True` (required to invoke this).
```
import torch
import torch.autograd.functional as F

def f(x):
    return x.sin_()

x = torch.tensor([1.2], requires_grad=True)
result, vjp = F.vjp(f, x.clone(), create_graph=True)

assert torch.isclose(vjp, x.cos())
# vjp is torch.tensor(1.0)
```
This produces a result that is different from what can be expected. As a user, my first interpretation of the above code is "compute the gradient of sin". 

Under the hood, the above is equivalent to doing the following:
```
x = torch.tensor([1.2], requires_grad=True)
y = x.clone()
vjp, = torch.autograd.grad(y.sin_(), y)
```
In this example, it's easier to see why `F.vjp` returns 1.0 -- because we modified the tensor being differentiated in-place.

## Action items

Should we warn the user about this in the documentation?

## Environment
master
